s3-streamer

```diff
@@ -0,0 +1,319 @@
+#!/usr/bin/python3
+
+# This file is part of Cockpit.
+
+# Copyright (C) 2022 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import argparse
+import codecs
+import json
+import locale
+import mimetypes
+import os
+import subprocess
+import tempfile
+import textwrap
+import time
+import urllib.parse
+
+from lib import s3
+from task import github
+
+
+class Destination:
+    def has(self, filename):
+        raise NotImplementedError
+
+    def write(self, filename, data):
+        raise NotImplementedError
+
+    def delete(self, filenames):
+        raise NotImplementedError
+
+
+class LocalDestination(Destination):
+    def __init__(self, directory):
+        self.directory = directory
+        os.makedirs(self.directory)
+
+    def path(self, filename):
+        return os.path.join(self.directory, filename)
+
+    def has(self, filename):
+        return os.path.exists(self.path(filename))
+
+    def write(self, filename, data):
+        print(f'Write {self.path(filename)}')
+        with open(self.path(filename), 'wb+') as file:
+            file.write(data)
+
+    def delete(self, filenames):
+        for filename in filenames:
+            print(f'Delete {self.path(filename)}')
+            os.unlink(self.path(filename))
+
+
+class S3Destination(Destination):
+    def __init__(self, directory):
+        self.directory = directory.rstrip('/') + '/'
+
+    def url(self, filename):
+        return urllib.parse.urlparse(self.directory + filename)
+
+    def has(self, filename):
+        raise NotImplementedError('use Index')
+
+    def write(self, filename, data):
+        headers = {
+            'Content-Type': mimetypes.guess_type(filename)[0] or 'text/plain',
+            s3.ACL: s3.PUBLIC
+        }
+        with s3.urlopen(self.url(filename), data=data, method='PUT', headers=headers) as response:
+            print(response.status, filename)
+
+    def delete(self, filenames):
+        # to do: multi-object delete API
+        for filename in filenames:
+            with s3.urlopen(self.url(filename), method='DELETE') as response:
+                print(response)
+
+
+class Index(Destination):
```

Comment:
Constantly maintaining an index.html, from dozens of parallel bots feels error prone and like unnecessary churn. We have both the S3 CLI and our test-results.db for querying available log URLs. I'd rather drop this.

Reply:
Each bot generates a unique subdirectory name and writes to it.  We never try to maintain an index.html for the toplevel.

All the same, I may indeed drop this code...

Reply:
I want to come up with a replacement before I drop this, though.  In particular, I want to send a static `index.html` which queries the REST API for the file list and generates an index on the client side.  That shouldn't be hard, but it's not on the critical path to landing this, so I'd rather do it later.